openshift/dynamic/client.py

```diff
@@ -66,18 +90,43 @@ class DynamicClient(object):
         the kubernetes API
     """
 
-    def __init__(self, client):
+    def __init__(self, client, cache_file=None):
         self.client = client
         self.configuration = client.configuration
+        default_cache_id = self.configuration.host
+        if PY3:
+            default_cache_id = default_cache_id.encode('utf-8')
+        self.__cache_file = cache_file or '/tmp/osrcp-{0}.json'.format(base64.b64encode(default_cache_id).decode('utf-8'))
+        self.__init_cache()
+
+    def __init_cache(self, refresh=False):
+        if refresh or not os.path.exists(self.__cache_file):
+            self.__cache = {}
+            refresh = True
+        else:
+            with open(self.__cache_file, 'r') as f:
+                self.__cache = json.load(f, cls=cache_decoder(self))
         self._load_server_info()
-        self.__resources = ResourceContainer(self.parse_api_groups())
+        self.__resources = ResourceContainer(self.parse_api_groups(), client=self)
+
+        if refresh:
+            self.__write_cache()
+
+    def __write_cache(self):
+        with open(self.__cache_file, 'w') as f:
+            json.dump(self.__cache, f, cls=CacheEncoder)
+
+    def invalidate_cache(self):
+        self.__init_cache(refresh=True)
 
     def _load_server_info(self):
-        self.__version = {'kubernetes': load_json(self.request('get', '/version'))}
-        try:
-            self.__version['openshift'] = load_json(self.request('get', '/version/openshift'))
-        except ApiException:
-            pass
+        if not self.__cache.get('version'):
+            self.__cache['version'] = {'kubernetes': load_json(self.request('get', '/version'))}
+            try:
+                self.__cache['version']['openshift'] = load_json(self.request('get', '/version/openshift'))
```

Comment:
I believe `/version/openshift` no longer exists, is this just to maintain legacy support?

Reply:
Yeah, I imagine this is usually run against older versions of openshift so I figured I would just leave it in.